crates/ruff_linter/src/rules/pylint/rules/repeated_keywords.rs

```diff
@@ -0,0 +1,93 @@
+use std::collections::HashSet;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{Expr, ExprDict, ExprStringLiteral, Keyword};
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for repeated keyword arguments passed to a function call
+///
+/// ## Why is this bad?
+/// Python does not allow for multiple values to be assigned to the same
+/// keyword argument in a single function call.
+///
+/// ## Example
+/// ```python
+/// func(1, 2, c=3, **{"c": 4})
+/// ```
+///
+/// Use instead:
+/// ```python
+/// func(1, 2, **{"c": 4})
+/// ```
+///
+/// ## References
+/// - [Python documentation: Argument](https://docs.python.org/3/glossary.html#term-argument)
+#[violation]
+pub struct RepeatedKeywords {
+    duplicate_keyword: String,
+}
+
+impl Violation for RepeatedKeywords {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let dupe = &self.duplicate_keyword;
+        format!("Repeated keyword argument: `{dupe}`")
+    }
+}
+
+type KeywordRecordFn<'a> = Box<dyn FnMut(&str, TextRange) + 'a>;
+
+fn generate_record_func<'a>(checker: &'a mut Checker) -> KeywordRecordFn<'a> {
+    // init some hash sets to be captured by the closure
+    let mut seen = HashSet::<String>::new();
+    let mut dupes = HashSet::<String>::new();
+
+    let inner = move |keyword: &str, range| {
+        // Add an error the first time we see the duplicate
+        if seen.contains(keyword) && !dupes.contains(keyword) {
+            dupes.insert(String::from(keyword));
+            checker.diagnostics.push(Diagnostic::new(
+                RepeatedKeywords {
+                    duplicate_keyword: keyword.into(),
+                },
+                range,
+            ));
+        } else {
+            seen.insert(String::from(keyword));
+        }
+    };
+
+    Box::new(inner)
+}
+
+pub(crate) fn repeated_keywords(checker: &mut Checker, keywords: &Vec<Keyword>) {
```

Comment:
Since you asked for nitpicks: it'd be more common to express this as `keywords: &[Keyword]` -- a slice, rather than a vector. You wouldn't need to change the call-site at all, but a slice is basically a more "permissive" type: you can always make a slice from a vector, but you can't go the other way around (if you have a slice like `&[Keyword]`, you can't pass that to a function that takes `&Vec<Keyword>`).